internal/controllers/configuration/kongadminapi_controller.go

```diff
@@ -0,0 +1,102 @@
+package configuration
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/samber/lo"
+	corev1 "k8s.io/api/core/v1"
+	discoveryv1 "k8s.io/api/discovery/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/adminapi"
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/util"
+)
+
+// KongAdminAPIServiceReconciler reconciles Kong Admin API Service Endpointslices
+// and notifies the provided notifier about those.
+type KongAdminAPIServiceReconciler struct {
+	client.Client
+
+	ServiceNN types.NamespacedName
+	Service   *corev1.Service
+
+	Log              logr.Logger
+	CacheSyncTimeout time.Duration
+
+	EndpointsNotifier EndpointsNotifier
+}
+
+type EndpointsNotifier interface {
+	Notify(addresses []string)
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *KongAdminAPIServiceReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	c, err := controller.New("KongAdminAPIEndpoints", mgr, controller.Options{
+		Reconciler: r,
+		LogConstructor: func(_ *reconcile.Request) logr.Logger {
+			return r.Log
+		},
+		CacheSyncTimeout: r.CacheSyncTimeout,
+	})
+	if err != nil {
+		return err
+	}
+
+	return c.Watch(
+		&source.Kind{Type: &discoveryv1.EndpointSlice{}},
+		&handler.EnqueueRequestForObject{},
+		predicate.NewPredicateFuncs(r.shouldReconcileEndpointSlice),
+	)
+}
+
+func (r *KongAdminAPIServiceReconciler) shouldReconcileEndpointSlice(obj client.Object) bool {
+	endpoints, ok := obj.(*discoveryv1.EndpointSlice)
+	if !ok {
+		return false
+	}
+
+	if !lo.ContainsBy(endpoints.OwnerReferences, func(ref metav1.OwnerReference) bool {
+		return ref.Kind == "Service" && ref.Name == r.ServiceNN.Name
+	}) {
+		return false
+	}
+
+	return true
+}
+
+//+kubebuilder:rbac:groups="discovery.k8s.io",resources=endpointslices,verbs=get;list;watch
+
+// Reconcile processes the watched objects.
+func (r *KongAdminAPIServiceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	var endpoints discoveryv1.EndpointSlice
+	if err := r.Get(ctx, req.NamespacedName, &endpoints); err != nil {
+		return ctrl.Result{}, err
+	}
+	r.Log.Info("reconciling EndpointSlice", "namespace", req.Namespace, "name", req.Name)
+
+	if !endpoints.DeletionTimestamp.IsZero() {
+		// TODO extract debug() and info log utilities to a shared package.
+		r.Log.V(util.DebugLevel).Info("EndpointSlice is being deleted",
+			"type", "EndpointSlice", "namespace", req.Namespace, "name", req.Name,
+		)
+		return ctrl.Result{}, nil
```

Comment:
Don't we need to tear down all the clients when the endpointslice is deleted?

Reply:
I just realized 1 thing. That wouldn't be necessary if we were using Endpoints, but since we're using EndpointSlice and there can be a lot of them for a single service, by default:

> By default, the control plane creates and manages EndpointSlices to have no more than 100 endpoints each. You can configure this with the --max-endpoints-per-slice [kube-controller-manager](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-controller-manager/) flag, up to a maximum of 1000.

This implementation is not correct whenever there's more than 1 EndpointSlice 😬 .

So there are a few options for this:

- introduce some state in `KongClient` and keep a mapping of endpoints to EndpointSlice, then whenever I'll get an update I send both the NN and list of URLs.
- watch `Service` and not `EndpointSlice` and send a whole set of endpoints at once: noisy (`Service` watch will be notified more than necessary) and will need to prepare a full list every time.
- keep state in the controller and only notify when addresses changed

Reply:
Good, that was my favourite too 🙃

Reply:
> I'd vote for this one - it keeps the notifying interface clean (KongClient side gets all clients that are to be used, no additional logic purely related to Kubernetes specifics there).

+1

Reply:
Fixed that by adding state cache to the controller.
Tested that with kind using

```
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4

kubeadmConfigPatches:
- |
  apiVersion: kubeadm.k8s.io/v1beta3
  kind: ClusterConfiguration
  controllerManager:
    extraArgs:
      max-endpoints-per-slice: "2"
```

to cause service to have multiple EndpointSlices backing it without spinning hundreds of gateway pods.